blog: Prerendering with `@preact/preset-vite` (#1174)
preactjs/preact-www: content/en/blog/prerendering-preset-vite.md

---
title: Prerendering with `@preact/preset-vite`
date: 2024-08-06
authors:
  - Ryan Christian
---

# Prerendering with Preset Vite

It's been a half-year since our prerendering plugin has somewhat quietly become available in `@preact/preset-vite`, so let's talk about it, our history, and the ecosystem at large

> Those who have been in our community for a while know how much we like prerendering; it was a first-class feature in Preact-CLI, then WMR, and now our Vite preset. When it's done right, it's a pain-free addition to the typical SPA that improves the user experience greatly and this plugin aims to facilitate just that.

Author: Thanks!

- Toggle it by flipping a Boolean in your config file
	- Because we're not a wrapper, and because you don't need to alter your source code in order to support it beyond some window checks, there's no lock-in whatsoever. If you decide to move away, or you want to do some testing on your output, all you need to do is flip a Boolean and you're back to a plain SPA with Vite.
	- As we've mentioned a few times, prerendering is at its best when it is as near to "drop-in" as possible and that includes being able to drop back out on a whim. It's important to us that you can go from an SPA to prerendering and back again with minimal effort.

## Final Notes

The Vite team would probably like us to mention that this plugin does introduce a tiny patch to the generated client code, and that they (the Vite team) don't necessarily endorse running the browser bundles in Node.

The patch in question is as follows:

```diff
// src/node/plugins/importAnalysisBuild.ts
-if (__VITE_IS_MODERN__ && deps && deps.length > 0) {,
+if (__VITE_IS_MODERN__ && deps && deps.length > 0 && typeof window !== 'undefined') {,
	 const links = document.getElementsByTagName('link')
	 ...
```

> As attempting to execute `document.getElementsByTagName` will error in Node where there is no `document` we simply add an additional condition to the preloader so that it makes no attempt to run in Node and that's it. Just the partial change of this one line.

Author: ```suggestion
As attempting to execute `document.getElementsByTagName` will error in Node where there is no `document`, we simply add an additional condition to the preloader so that it makes no attempt to run in Node, and that's it. Just the partial change of this one line.
```

> However, we are very, very happy with this level of risk and have been suing it heavily for some time now without any issue, but, we are somewhat using the tool beyond what it was intended for and it's something we want to disclose.

Author: ```suggestion
We are very, very happy with this level of risk and have been using it heavily for some time now without any issue, but, this is somewhat using the tool beyond what it was intended for and it's something we want to disclose.
```

1. Multiple Builds
	- Separate client/server builds, often separate entry points too
	- Less isomorphic, different branches in your app for different environments
2. Frameworks / Vite Wrappers
	- No longer using Vite directly but an abstraction
	- Some amount of buy-in/lock-in
	- Support matrix for different Vite config options, plugins, etc. can be complicated and less than clear

While these solutions absolutely have their merits and places in the ecosystem, neither felt as great as they could be for our ecosystem, given our historic offerings in this area. The "best case scenario" DX was often sacrificed for more complex or specific needs -- which is a completely valid trade off.

For drop-in prerendering, however, we thought we could provide something a bit different to the existing options, or at least something a bit more familiar to our users.

## Implementation in `@preact/preset-vite`

> Year later, we were still quite in love with the simplicity and extensibility of WMR's prerendering and felt it was sorely missing from our Vite preset, so we looked to port it over with a few minor adjustment to fix the qualms we had. A little bit of work later and voila, prerendering via a plugin!

Reviewer: Should this be `A year later,...`?

Author: `Years`, thanks!

```suggestion
Years later, we were still quite in love with the simplicity and extensibility of WMR's prerendering and felt it was sorely missing from our Vite preset, so we looked to port it over with a few minor adjustment to fix the qualms we had. A little bit of work later and voila, prerendering via a plugin!
```

```jsx
// src/use-fetch.js
import { useState } from "preact/hooks";

const cache = new Map();

async function load(url) {
	const res = await fetch(url);
	if (res.ok) return await res.text();
	throw new Error(`Failed to fetch ${url}!`);
}

// Simple suspense-based fetch mechanism with caching
export function useFetch(url) {
	const [_, update] = useState({});

	let data = cache.get(url);
	if (!data) {
		data = load(url);
		cache.set(url, data);
		data.then(
			(res) => update((data.res = res)),
			(err) => update((data.err = err)),
		);
	}

	if (data.res) return data.res;
	if (data.err) throw data.err;
	throw data;
}
```

```jsx
// src/index.jsx
import { hydrate, prerender as ssr } from 'preact-iso';
import { useFetch } from './use-fetch.js';

function App() {
    return (
	    <div>
		    <Suspense fallback={<p>Loading...</p>}>
			    <Article />
		    </Suspense>
		</div>
    );
}

function Article() {
	const data = useFetch("/my-local-article.txt");
	return <p>{data}</p>;
}

if (typeof window !== 'undefined') {
	hydrate(<App />, document.getElementById('app'));
}

export async function prerender(data) {
    return await ssr(<App {...data} />)
}
```

> ### Using `globalThis` to pass data around

Reviewer: I like the later example of having a global cache to write to, but I'm a bit unsure about passing other data. I'm worried that this could lead to race conditions given that we also support async rendering in RTS now.

Author: There might be a better example in there. This sort of thing is something I've pointed out to people enough times that I think I want *something* there, as it's fairly reasonable to need to pass data back up (and out of the tree). Will have a think.

Shouldn't be any race conditions AFAIK; component tree will have to have fully rendered before the RTS pass completes? That's not to say you can't cause issues, but... shouldn't be anything inherent to the pattern?
